File: src/loom/validation.py
"""Read-only validation: report inconsistencies between DB and filesystem.

Validate never mutates anything (not the DB, not the markdown). For
fixes, point users at :func:`loom.rebuild.rebuild` or
:func:`loom.rebuild.sync_one`.

Cycle detection is intentionally out of scope here — Phase 4's
``deps`` module owns that.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

from .errors import InvalidQualifiedId, ValidationError
from .ids import qid_from_path
from .index import Index
from .scan import hash_file_bytes, walk_md_files
from .storage import load

# Stable kind constants for JSON consumers.
KIND_BROKEN_DEP = "broken_dep"
KIND_DRIFT = "drift"
KIND_PATH_ID_MISMATCH = "path_id_mismatch"
KIND_TYPE_MISMATCH = "type_mismatch"
KIND_PARSE_ERROR = "parse_error"
KIND_STRAY_FILE = "stray_file"
KIND_MISSING_FIELD = "missing_field"
KIND_NOT_INDEXED = "not_indexed"
KIND_ORPHAN_INDEX = "orphan_index"


@dataclass(frozen=True, slots=True)
class ValidationIssue:
    kind: str
    qualified_id: str | None
    file_path: str | None
    message: str

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def validate(root: Path) -> list[ValidationIssue]:
    """Walk the filesystem and the index, returning every inconsistency."""
    issues: list[ValidationIssue] = []
    idx = Index(root)

    indexed_qids = set(idx.all_qualified_ids())
    indexed_records = {qid: idx.get(qid) for qid in indexed_qids}

    seen_qids: set[str] = set()

    for path in walk_md_files(root):
        rel = str(path.relative_to(root))
        try:
            qid, _archived = qid_from_path(path, root)
        except InvalidQualifiedId as e:
            issues.append(ValidationIssue(KIND_STRAY_FILE, None, rel, str(e)))
            continue

        try:
            fm, _body = load(path)
        except ValidationError as e:
            issues.append(ValidationIssue(KIND_PARSE_ERROR, str(qid), rel, str(e)))
            continue

        seen_qids.add(str(qid))

        # Frontmatter qualified_id / id mismatches with the path.
        if (fm_qid := fm.get("qualified_id")) is not None and str(fm_qid) != str(qid):
            issues.append(
                ValidationIssue(
                    KIND_PATH_ID_MISMATCH,
                    str(qid),
                    rel,
                    f"frontmatter qualified_id {str(fm_qid)!r} != path-derived {str(qid)!r}",
                )
            )
        if (fm_local := fm.get("id")) is not None and str(fm_local) != qid.local_id:
            issues.append(
                ValidationIssue(
                    KIND_PATH_ID_MISMATCH,
                    str(qid),
                    rel,
                    f"frontmatter id {str(fm_local)!r} != path-derived {qid.local_id!r}",
                )
            )

        # Type mismatch.
        if (fm_type := fm.get("type")) is not None and str(fm_type) != qid.type.value:
            issues.append(
                ValidationIssue(
                    KIND_TYPE_MISMATCH,
                    str(qid),
                    rel,
                    f"frontmatter type {str(fm_type)!r} != path-derived {qid.type.value!r}",
                )
            )

        # Missing required fields.
        for field_name in ("title", "created_at", "updated_at"):
            if fm.get(field_name) is None:
                issues.append(
                    ValidationIssue(
                        KIND_MISSING_FIELD,
                        str(qid),
                        rel,
                        f"missing required field {field_name!r}",
                    )
                )
        if qid.type.value != "project" and fm.get("status") is None:
            issues.append(
                ValidationIssue(
                    KIND_MISSING_FIELD,
                    str(qid),
                    rel,
                    "missing required field 'status'",
                )
            )

        # Drift / not-indexed.
        rec = indexed_records.get(str(qid))
        if rec is None:
            issues.append(
                ValidationIssue(
                    KIND_NOT_INDEXED,
                    str(qid),
                    rel,
                    "file exists on disk but is not in the index; run `loom rebuild`",
                )
            )
        else:
            on_disk = hash_file_bytes(path)
            if rec.body_hash != on_disk:
                issues.append(
                    ValidationIssue(
                        KIND_DRIFT,
                        str(qid),
                        rel,
                        "on-disk content differs from indexed body_hash",
                    )
                )

        # Broken deps: read from frontmatter (source of truth) so we still
        # surface dependencies whose targets were missing at index time and
        # therefore never landed in the dependencies table.
        for dep in fm.get("depends_on") or ():
            dep_str = str(dep)
            if dep_str not in indexed_qids:
                issues.append(
                    ValidationIssue(
                        KIND_BROKEN_DEP,
                        str(qid),
                        rel,
                        f"depends on missing item {dep_str}",
                    )
                )

    # Index entries with no corresponding file.
    for qid in sorted(indexed_qids - seen_qids):
        rec = indexed_records[qid]
        issues.append(
            ValidationIssue(
                KIND_ORPHAN_INDEX,
                qid,
                rec.file_path if rec else None,
                "indexed but no matching file on disk; run `loom rebuild`",
            )
        )

    return issues

File: web/lib/validation.ts
/**
 * Read-only validation: report inconsistencies between DB and filesystem.
 *
 * validate() never mutates anything (not the DB, not the markdown).
 * For fixes, use rebuild() or syncOne() from rebuild.ts.
 *
 * Mirrors src/loom/validation.py.
 */

import { load } from "./storage";
import { qid_from_path } from "./ids";
import { Index, initDb, DB_FILENAME } from "./index";
import { walkMdFiles, hashFileBytes } from "./scan";
import { InvalidQualifiedId, ValidationError } from "./errors";
import * as path from "path";
import * as fs from "fs";

// ---------------------------------------------------------------------------
// Stable KIND constants — part of the JSON consumer contract.
// ---------------------------------------------------------------------------

export const KIND_BROKEN_DEP = "broken_dep";
export const KIND_DRIFT = "drift";
export const KIND_PATH_ID_MISMATCH = "path_id_mismatch";
export const KIND_TYPE_MISMATCH = "type_mismatch";
export const KIND_PARSE_ERROR = "parse_error";
export const KIND_STRAY_FILE = "stray_file";
export const KIND_MISSING_FIELD = "missing_field";
export const KIND_NOT_INDEXED = "not_indexed";
export const KIND_ORPHAN_INDEX = "orphan_index";

// ---------------------------------------------------------------------------
// ValidationIssue
// ---------------------------------------------------------------------------

export interface ValidationIssue {
  kind: string;
  qualified_id: string | null;
  file_path: string | null;
  message: string;
}

// ---------------------------------------------------------------------------
// validate
// ---------------------------------------------------------------------------

/**
 * Walk the filesystem and the index, returning every inconsistency found.
 * Never mutates the DB or any file.
 */
export async function validate(root: string): Promise<ValidationIssue[]> {
  const issues: ValidationIssue[] = [];

  // Ensure DB exists so Index can open it.
  const dbPath = path.join(root, DB_FILENAME);
  if (!fs.existsSync(dbPath)) {
    initDb(dbPath);
  }

  const idx = new Index(root);
  const indexedQids = new Set(idx.allQualifiedIds());
  const indexedRecords = new Map<string, ReturnType<typeof idx.get>>();
  for (const qid of indexedQids) {
    indexedRecords.set(qid, idx.get(qid));
  }

  const seenQids = new Set<string>();

  for (const filePath of walkMdFiles(root)) {
    const rel = path.relative(root, filePath);
    let qid: ReturnType<typeof qid_from_path>[0];
    try {
      [qid] = qid_from_path(filePath, root);
    } catch (e) {
      if (e instanceof InvalidQualifiedId) {
        issues.push({ kind: KIND_STRAY_FILE, qualified_id: null, file_path: rel, message: String(e) });
      } else {
        issues.push({ kind: KIND_STRAY_FILE, qualified_id: null, file_path: rel, message: String(e) });
      }
      continue;
    }

    let fm: Record<string, unknown>;
    try {
      [fm] = load(filePath);
    } catch (e) {
      if (e instanceof ValidationError) {
        issues.push({ kind: KIND_PARSE_ERROR, qualified_id: qid.toString(), file_path: rel, message: String(e) });
      } else {
        issues.push({ kind: KIND_PARSE_ERROR, qualified_id: qid.toString(), file_path: rel, message: String(e) });
      }
      continue;
    }

    seenQids.add(qid.toString());

    // Frontmatter qualified_id / id mismatches with path.
    const fmQid = fm["qualified_id"];
    if (fmQid !== undefined && fmQid !== null && String(fmQid) !== qid.toString()) {
      issues.push({
        kind: KIND_PATH_ID_MISMATCH,
        qualified_id: qid.toString(),
        file_path: rel,
        message: `frontmatter qualified_id ${JSON.stringify(String(fmQid))} != path-derived ${JSON.stringify(qid.toString())}`,
      });
    }
    const fmLocal = fm["id"];
    if (fmLocal !== undefined && fmLocal !== null && String(fmLocal) !== qid.localId) {
      issues.push({
        kind: KIND_PATH_ID_MISMATCH,
        qualified_id: qid.toString(),
        file_path: rel,
        message: `frontmatter id ${JSON.stringify(String(fmLocal))} != path-derived ${JSON.stringify(qid.localId)}`,
      });
    }

    // Type mismatch.
    const fmType = fm["type"];
    if (fmType !== undefined && fmType !== null && String(fmType) !== qid.type) {
      issues.push({
        kind: KIND_TYPE_MISMATCH,
        qualified_id: qid.toString(),
        file_path: rel,
        message: `frontmatter type ${JSON.stringify(String(fmType))} != path-derived ${JSON.stringify(qid.type)}`,
      });
    }

    // Missing required fields.
    for (const fieldName of ["title", "created_at", "updated_at"]) {
      if (fm[fieldName] === undefined || fm[fieldName] === null) {
        issues.push({
          kind: KIND_MISSING_FIELD,
          qualified_id: qid.toString(),
          file_path: rel,
          message: `missing required field ${JSON.stringify(fieldName)}`,
        });
      }
    }
    if (qid.type !== "project" && (fm["status"] === undefined || fm["status"] === null)) {
      issues.push({
        kind: KIND_MISSING_FIELD,
        qualified_id: qid.toString(),
        file_path: rel,
        message: "missing required field 'status'",
      });
    }

    // Drift / not-indexed.
    const rec = indexedRecords.get(qid.toString()) ?? null;
    if (rec === null) {
      issues.push({
        kind: KIND_NOT_INDEXED,
        qualified_id: qid.toString(),
        file_path: rel,
        message: "file exists on disk but is not in the index; run `loom rebuild`",
      });
    } else {
      const onDisk = await hashFileBytes(filePath);
      if (rec.body_hash !== onDisk) {
        issues.push({
          kind: KIND_DRIFT,
          qualified_id: qid.toString(),
          file_path: rel,
          message: "on-disk content differs from indexed body_hash",
        });
      }
    }

    // Broken deps: read from frontmatter (source of truth) so we still
    // surface deps whose targets were missing at index time.
    const rawDeps = fm["depends_on"];
    const deps = Array.isArray(rawDeps) ? rawDeps.map(String) : [];
    for (const dep of deps) {
      if (!indexedQids.has(dep)) {
        issues.push({
          kind: KIND_BROKEN_DEP,
          qualified_id: qid.toString(),
          file_path: rel,
          message: `depends on missing item ${dep}`,
        });
      }
    }
  }

  // Index entries with no corresponding file.
  for (const qid of [...indexedQids].sort()) {
    if (!seenQids.has(qid)) {
      const rec = indexedRecords.get(qid) ?? null;
      issues.push({
        kind: KIND_ORPHAN_INDEX,
        qualified_id: qid,
        file_path: rec ? rec.file_path : null,
        message: "indexed but no matching file on disk; run `loom rebuild`",
      });
    }
  }

  return issues;
}
